### PR TITLE
[WIP] additionnal lib/cleanup for French language to improve quality of inputs

### PR DESCRIPTION
**Pull request by CapitainFlam**
<H1>better french input for better french output ?</h1>

Hi ! 
⚠️ Firsts commits in GitHub, first pull request ever in a repo, so please neither shoot or shout at me 😅  ⚠️ 

This commit comes from different discussion, that I propose you to have a look to understand the context. 

- https://github.com/common-voice/commonvoice-fr/issues/21 _(...This issue should be partially solved with this pull)_
- https://github.com/common-voice/CorporaCreator/pull/87 _(...This is inspirational commit that I took some ideas from)_

and it is somehow connected to : 

- https://github.com/common-voice/commonvoice-fr/issues/20
- https://commonvoice.mozilla.org/fr/criteria
- https://discourse.mozilla.org/t/rfc-criteres-dacceptation-pour-la-validation-denregistrements/73063/12 
- https://discourse.mozilla.org/t/proportion-de-noms-propres-a-consonnance-etrangere/72887/26
- https://github.com/common-voice/common-voice/pull/3786
- https://github.com/common-voice/common-voice/issues/3792
- https://github.com/common-voice/cv-sentence-extractor/issues/92

**TL;DR: 
I'm trying to improve input quality for french sentence collector, to avoid 
**_A/_** garbage IN - garbage OUT, and 
_B/_ avoid to remove in the end (in the [CorporaCreator](https://github.com/common-voice/CorporaCreator)) the garbage that went through collecting, recording, review... to finally being dropped and not being included in the final release for training batchs.** Better to remove it as soon as sentence collector.

<H1>What have you done ? 😱 </h1>

To do so, I duplicated a EN to FR file in sentence collector, and modified it according to a previous job made by Nicolas Panel for CorporaCreator (and sadly not commited). 

According to discussion (links [here](https://github.com/common-voice/commonvoice-fr/issues/21#issuecomment-1242611833)... Did you follow it from the list above ?!), it's recommanded that I create a WIP pull request, to allow everyone to comment, throw tomatos and/or additionnal commits to it. 

footnote : it can be hard (...it IS hard !!!) to understand REGEX (REGular EXpressions). Do not hesitate to catch up with https://regex101.com/ to understand and test it.

### CODE REVIEW

File: server/lib/cleanup/languages/fr.js [hunk 1]
```diff
@@ -49,9 +51,134 @@ function clean(sentences) {
...
+      .replace(/(^|\s)RSA(\s|\.|,|\?|!|$)/g, ' Revenu de Solidarité Active ')
+      .replace(/(^|\s)RSA(\s|\.|,|\?|!|$)/g, ' Revenu de Solidarité Active ')
+      .replace(/(^|\s)RSI(\s|\.|,|\?|!|$)/g, ' Régime Social des Indépendants ')
+      .replace(/(^|\s)RTE(\s|\.|,|\?|!|$)/g, ' Réseau de Transport d\'Électricité ')
+      .replace(/(^|\s)SNCF(\s|\.|,|\?|!|$)/g, ' Société Nationale des Chemins de Fer ')
+      .replace(/(^|\s)TGV(\s|\.|,|\?|!|$)/g, ' Train à Grande Vitesse ')
+      .replace(/(^|\s)TVA(\s|\.|,|\?|!|$)/g, ' Taxe sur la Valeur Ajoutée ')
+      .replace(/(^|\s)UDI(\s|\.|,|\?|!|$)/g, ' Union des Démocrates Indépendants ')
+      .replace(/(^|\s)UMP(\s|\.|,|\?|!|$)/g, ' Union pour un Mouvement Populaire ')
+      .replace(/(^|\s)USA(\s|\.|,|\?|!|$)/g, ' Etats Unis d\'Amérique ')
```
Comment (drzraf) [inline server/lib/cleanup/languages/fr.js:167]: Side note: Acronyms are a strange case : They are actually spelled as distincts letters (except for NUPES).

But we **don't** want "STT" to process `R.S.A.` as `erre et ça` **but we currently have nothing to train it that way**.

Should we expect a specific future dictionary or just not strip some of them in the first place (but remove the dots)?

File: server/lib/cleanup/languages/fr.js [hunk 2]
```diff
@@ -49,9 +51,134 @@ function clean(sentences) {
...
+	  .replace(/(^|\s)XIII(\s|\.|,|\?|!|$)/g, ' treize ')
+	  .replace(/(^|\s)(XIIII|XIV)(\s|\.|,|\?|!|$)/g, ' quatorze ')
+	  .replace(/(^|\s)XV(\s|\.|,|\?|!|$)/g, ' quinze ')
+	  .replace(/(^|\s)XVI(\s|\.|,|\?|!|$)/g, ' seize ')
+	  .replace(/(^|\s)XVII(\s|\.|,|\?|!|$)/g, ' dix-sept ')
+	  .replace(/(^|\s)XVIII(\s|\.|,|\?|!|$)/g, ' dix-huit ')
+	  .replace(/(^|\s)(XIX|XVIIII)(\s|\.|,|\?|!|$)/g, ' dix-neuf ')
+	  .replace(/(^|\s)XX(\s|\.|,|\?|!|$)/g, ' vingt ')
+ 	  .replace(/(^|\s)XXI(\s|\.|,|\?|!|$)/g, ' vingt-et-un ')
+ 	  .replace(/(^|\s)XXII(\s|\.|,|\?|!|$)/g, ' vingt-deux ')
```
Comment (MichaelKohler) [inline server/lib/cleanup/languages/fr.js:120]: I think we can remove all of the roman numerals, as they are not allowed from the beginning (see the abbreviation pattern in the validation file).

File: server/lib/cleanup/languages/fr.js [hunk 3]
```diff
@@ -49,9 +51,134 @@ function clean(sentences) {
...
+	  .replace(/(^|\s)2n?ds?(\s|\.|,|\?|!|$)/g, ' second ')
+	  .replace(/(^|\s)2n?des?(\s|\.|,|\?|!|$)/g, ' seconde ')
+	  .replace(/(^|\s)3i?(e|è)me?s?(\s|\.|,|\?|!|$)/g, ' troisième ')
+	  .replace(/(^|\s)4i?(e|è)me?s?(\s|\.|,|\?|!|$)/g, ' quatrième ')
+	  .replace(/(^|\s)5i?(e|è)me?s?(\s|\.|,|\?|!|$)/g, ' cinquième ')
+	  .replace(/(^|\s)6i?(e|è)me?s?(\s|\.|,|\?|!|$)/g, ' sixième ')
+	  .replace(/(^|\s)7i?(e|è)me?s?(\s|\.|,|\?|!|$)/g, ' septième ')
+	  .replace(/(^|\s)8i?(e|è)me?s?(\s|\.|,|\?|!|$)/g, ' huitième ')
+	  .replace(/(^|\s)9i?(e|è)me?s?(\s|\.|,|\?|!|$)/g, ' neuvième ')
+	  .replace(/(^|\s)10i?(e|è)me?s?(\s|\.|,|\?|!|$)/g, ' dixième ')
```
Comment (MichaelKohler) [inline server/lib/cleanup/languages/fr.js:135]: These can be removed, as numbers are not allowed anyway in the Sentence Collector.

File: server/lib/cleanup/languages/fr.js (hunk 1 above)
Comment (MichaelKohler) [inline server/lib/cleanup/languages/fr.js:167]: These can be removed, as abbreviations are not allowed in the validation file.

File: server/lib/cleanup/languages/fr.js [hunk 4]
```diff
@@ -49,9 +51,134 @@ function clean(sentences) {
...
+      .replace(/(^|\s)RTE(\s|\.|,|\?|!|$)/g, ' Réseau de Transport d\'Électricité ')
+      .replace(/(^|\s)SNCF(\s|\.|,|\?|!|$)/g, ' Société Nationale des Chemins de Fer ')
+      .replace(/(^|\s)TGV(\s|\.|,|\?|!|$)/g, ' Train à Grande Vitesse ')
+      .replace(/(^|\s)TVA(\s|\.|,|\?|!|$)/g, ' Taxe sur la Valeur Ajoutée ')
+      .replace(/(^|\s)UDI(\s|\.|,|\?|!|$)/g, ' Union des Démocrates Indépendants ')
+      .replace(/(^|\s)UMP(\s|\.|,|\?|!|$)/g, ' Union pour un Mouvement Populaire ')
+      .replace(/(^|\s)USA(\s|\.|,|\?|!|$)/g, ' Etats Unis d\'Amérique ')
+
+      //replace fraction 1/2 => '1 sur 2'
+     ¨.replace(/(^| )(\d+)(\s)?(\/)(\s)?(\d+)(\s|\.|,|\?|!|$)/g, '$2 sur $6')
```
Comment (MichaelKohler) [inline server/lib/cleanup/languages/fr.js:170]: Numbers are not allowed by the validation file.

File: server/lib/cleanup/languages/fr.js [hunk 5]
```diff
@@ -49,9 +51,134 @@ function clean(sentences) {
...
+
+	  //dates, digits and numbers fr-FR cleanup
+	  //todo : CONVERT TO TEXT instead of removing it
+	  .replace((^|\s)\d{1,2}\/\d{1,2}\/(\d{2}[^\d]|\d{4})(\s|$), ' ') //date format dd/mm/yy ou dd/mm/yyyy
+	  .replace((^|\s)\d{1,2}\/(\d{2}[^\d]|\d{4})(\s|$), ' ') //date format mm/yy ou mm/yyyy
```
Comment (MichaelKohler) [inline server/lib/cleanup/languages/fr.js:175]: Numbers are not allowed by the validation file.

File: server/lib/validation/languages/fr.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+// Maximum of words allowed per sentence to keep recordings in a manageable duration.
+const MAX_WORDS = 14;
+
+const INVALIDATIONS = [{
+  fn: (sentence) => {
+    const words = tokenizeWords(sentence);
+    return words.length < MIN_WORDS || words.length > MAX_WORDS;
+  },
+  error: `Number of words must be between ${MIN_WORDS} and ${MAX_WORDS} (inclusive) - Le nombre de mot doit être entre ${MIN_WORDS} et ${MAX_WORDS} (inclus)`,
```
Comment (MichaelKohler) [inline server/lib/validation/languages/fr.js:14]: I probably would leave only the French error message in here (goes for all of them). I don't particularly see a benefit of having both languages shown in the UI.

**Review by MichaelKohler — CHANGES_REQUESTED** [review]
> ⚠️ Firsts commits in GitHub, first pull request ever in a repo, so please neither shoot or shout at me 😅 ⚠️

No worries, you're doing fine :) Thanks for your contributions!

I have a few comments, but generally I think it's a good thing to update these files.

**Comment by MichaelKohler** [comment]
Also note https://discourse.mozilla.org/t/sentence-collector-cleanup-before-export-vs-cleanup-on-upload/105411, though that probably doesn't matter too much for this here :)

File: server/lib/cleanup/languages/fr.js [hunk 7]
```diff
@@ -49,9 +51,134 @@ function clean(sentences) {
...
+	  //based on common-voice/CorporaCreator#87
+
+	  //Jean-Paul II
+	  .replace(/Jean-Paul II|Jean Paul II/g, 'Jean-Paul deux')
+	  
+	  //abrevation fr-FR cleanup
+	  //based on common-voice/CorporaCreator#87
+	  .replace(/(^|\s|\w)\/an(\s|\.|,|\?|!|$)/g, '$1 par an ')
+	  .replace(/(^|\s)km(\s|\.|,|\?|!|$)/g, ' kilomètres ')
+	  .replace(/%, ' pourcent ')
```
Comment (MichaelKohler) [inline server/lib/cleanup/languages/fr.js:64]:
```suggestion
	  .replace(/%/, ' pourcent ')
```

File: server/lib/cleanup/languages/fr.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (CapitainFlam) [reply server/lib/cleanup/languages/fr.js:170]: oh shoot ! you're right ! 
To solve it, we shall test numbers, and see "if it's in date [1..31]/[1..12] range, i shall be a date stamp, otherwise it's a fraction ?!"
Sometimes, only context will show us that "mettre 1/4 de litre de lait" will go to "mettre un avril de litre de lait"... But it shall be more readable than "mettre de litre de lait".

File: server/lib/cleanup/languages/fr.js [hunk 8]
```diff
@@ -10,7 +10,9 @@ function sortSentences(sentences) {
 function clean(sentences) {
   return sentences.map((sentence) => {
     return sentence
-      // no space after opening '(' or '['
+	  //caracters and space cleanup
+	  
+	  // no space after opening '(' or '['
```
[the remark this answers is not in the document]
Comment (CapitainFlam) [reply server/lib/cleanup/languages/fr.js:15]: Yep. Wrong configuration of Notepad++ on my side (Tab instead of SpaceSpaceSpaceSpace) 😅 
Will correct it next time.

File: server/lib/cleanup/languages/fr.js [hunk 9]
```diff
@@ -49,9 +51,134 @@ function clean(sentences) {
       .replace(/\?(?!\s+)/g, '? ')
       .replace(/!(?!\s+)/g, '! ')
 
-      // Final normalization of spaces
-      .replace(/\s+/g, ' ')
+	  //special names and places cleanup  	  
+	  //based on common-voice/CorporaCreator#87
+
+	  //Jean-Paul II
+	  .replace(/Jean-Paul II|Jean Paul II/g, 'Jean-Paul deux')
```
[the remark this answers is not in the document]
Comment (CapitainFlam) [reply server/lib/cleanup/languages/fr.js:58]: It was just the work of Nicolas. ...I think we can remove it. (It was based on the dump file of the CorporaCreator errors, it could explain this 'only 1 entry' list)

File: server/lib/cleanup/languages/fr.js (hunk 1 above)
Comment (CapitainFlam) [reply server/lib/cleanup/languages/fr.js:167]: Well, I think I understand wrongly how it works. Let's solve this https://github.com/common-voice/sentence-collector/issues/636 issue first 😸

File: server/lib/cleanup/languages/fr.js [hunk 10]
```diff
@@ -49,9 +51,134 @@ function clean(sentences) {
...
+	  .replace(/(^|\s)XIII(e|è)(me)? s.(\s|\.|,|\?|!|$)/g, ' treizième siècle ')
+	  .replace(/(^|\s)(XIIII|XIV)(e|è)(me)? s.(\s|\.|,|\?|!|$)/g, ' quatorzième siècle ')
+	  .replace(/(^|\s)XV(e|è)(me)? s.(\s|\.|,|\?|!|$)/g, ' quinzième siècle ')
+	  .replace(/(^|\s)XVI(e|è)(me)? s.(\s|\.|,|\?|!|$)/g, ' seixième siècle ')
+	  .replace(/(^|\s)XVII(e|è)(me)? s.(\s|\.|,|\?|!|$)/g, ' dix-septième siècle ')
+	  .replace(/(^|\s)XVIII(e|è)(me)? s.(\s|\.|,|\?|!|$)/g, ' dix-huitième siècle ')
+	  .replace(/(^|\s)(XIX|XVIIII)(e|è)(me)? s.(\s|\.|,|\?|!|$)/g, ' dix_neuvième siècle ')
+	  .replace(/(^|\s)XX(e|è)(me)? s.(\s|\.|,|\?|!|$)/g, ' vingtième siècle ')
+	  .replace(/(^|\s)XXI(e|è)(me)? s.(\s|\.|,|\?|!|$)/g, ' vingt-et-unième siècle ')
+	  .replace(/(^|\s)XXII(e|è)(me)? s.(\s|\.|,|\?|!|$)/g, ' vingt-deuxième siècle ')
```
[the remark this answers is not in the document]
Comment (CapitainFlam) [reply server/lib/cleanup/languages/fr.js]: well, 
### First, 
I have my regex to work with : 
-- kudos to [gregory-r](https://stackoverflow.com/users/2872366/gregory-r) in [this post in stackoverflow](https://stackoverflow.com/a/49946096/20003021), licence [CC BY-SA](https://stackoverflow.com/help/licensing)

```javascript
function deromanize(roman) {
  var r = 0;
  // regular expressions to check if valid Roman Number.
  if (!/^M*(?:D?C{0,3}|C[MD])(?:L?X{0,3}|X[CL])(?:V?I{0,3}|I[XV])$/.test(roman))
    throw new Error('Invalid Roman Numeral.');

  roman.replace(/[MDLV]|C[MD]?|X[CL]?|I[XV]?/g, function(i) {
    r += {M:1000, CM:900, D:500, CD:400, C:100, XC:90, L:50, XL:40, X:10, IX:9, V:5, IV:4, I:1}[i]; 
  });

  return r;
}
```
So, I'll put it here to not forget it, because : 

### Second, 
I don't have a clue of how to implement a function in JS ! (I know Pascal and VB ✌️ , but never learned JS 😭... I have to jump in the rabbit hole 🕳️ 🐇  (it is [this way](https://www.youtube.com/watch?v=6IDT3MpSCKI)) )

...To however follow, either the white rabbit, or [crumbs](https://dictionary.cambridge.org/fr/dictionnaire/anglais/crumb), including future me, well, this link
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace#specifying_a_function_as_the_replacement 
could be a good start to look at.

/me : jumps in the 🕳️

File: server/lib/cleanup/languages/fr.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (CapitainFlam) [reply server/lib/cleanup/languages/fr.js]: I actually don't know how to import external library, I think it's a bit out of my league right now.

File: server/lib/cleanup/languages/fr.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (CapitainFlam) [reply server/lib/cleanup/languages/fr.js]: Est ce que le commit https://github.com/common-voice/sentence-collector/pull/635/commits/aef428430ae01712fb491e09dae157b9b4b62c3b répond au problème ?

File: server/lib/validation/languages/fr.js [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+}, {
+  regex: /[0-9]+/,
+  error: 'Sentence should not contain numbers - les phrases ne doivent pas contenir de nombres',
+}, {
+  regex: /[<>+*#@%^[\]()/]/,
+  error: 'Sentence should not contain symbols - les phrases de doivent pas contenir de symboles \(\*, \#, \(, etc\)',
```
[the remark this answers is not in the document]
Comment (CapitainFlam) [reply server/lib/validation/languages/fr.js]: commit https://github.com/common-voice/sentence-collector/pull/635/commits/18fdf37f6554dc015b19c82cac1b5627a82e766d

File: server/lib/validation/languages/fr.js [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+}, {
+  // Any words consisting of uppercase letters or uppercase letters with a period
+  // inbetween are considered abbreviations or acronyms.
+  // This currently also matches fooBAR but we most probably don't want that either
+  // as users wouldn't know how to pronounce the uppercase letters.
+  regex: /[A-Z]{2,}|[A-Z]+\.*[A-Z]+/,
+  error: 'Sentence should not contain abbreviations - Les phrases ne doivent pas contenir des abréviations ou sigles',
+}];
```
[the remark this answers is not in the document]
Comment (CapitainFlam) [reply server/lib/validation/languages/fr.js:33]: commit https://github.com/common-voice/sentence-collector/pull/635/commits/1a0cff5772ac2015ed8e2e90660069913fd75fac

File: server/lib/cleanup/languages/fr.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (CapitainFlam) [reply server/lib/cleanup/languages/fr.js:169]: Let's first wait for issue https://github.com/common-voice/sentence-collector/issues/636 resolution, 'coz I don't want to work for nothing :-/

File: server/lib/cleanup/languages/fr.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (CapitainFlam) [reply server/lib/cleanup/languages/fr.js]: Indeed, I don't understand that, collecting sentences, we can NOT "correct and clean" before "validation".

no problem, I'll wait for the issue  https://github.com/common-voice/sentence-collector/issues/636 😸

File: server/lib/validation/languages/fr.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (CapitainFlam) [reply server/lib/validation/languages/fr.js:33]: BTW, sources files like https://github.com/common-voice/commonvoice-fr/blob/master/CommonVoice-Data/data/debats-assemblee-nationale/20130718093000000.txt are using french apostrophe ?! 
(it only shows in "ui-monospace" code font, not the usual "Segoe UI" font. 
```
line 2 : J’appelle maintenant, dans le texte de la commission, les articles du projet de loi.
if I change it myself : 
line 2 : J'appelle maintenant, dans le texte de la commission, les articles du projet de loi.
```

[the remark this answers is not in the document]
Comment (CapitainFlam) [reply server/lib/validation/languages/fr.js:33]: so... commit https://github.com/common-voice/sentence-collector/pull/635/commits/8d7d97340fb5e98c47228a8fd709d46d1968b7f1

File: server/lib/cleanup/languages/fr.js (hunk 4 above)
Comment (CapitainFlam) [reply server/lib/cleanup/languages/fr.js:169]: Shouldn't we keep it for "future-proof rules" and/or "bulk load rules if you don't have yours" ? 
Let's discuss it here : https://discourse.mozilla.org/t/sentence-collector-cleanup-before-export-vs-cleanup-on-upload/105411/15

**Comment by drzraf** [comment]
Any hope to get this in, in one shape or another?

**Comment by MichaelKohler** [comment]
The Sentence Collector has now been integrated into the Common Voice platform. Therefore I'm archiving this project here. The validation files now live here and I'm sure it would still benefit from the validation rules being added there: https://github.com/common-voice/common-voice/tree/main/server/src/core/sentences/validation. Unfortunately moving this PR over there is way harder than manually recreating it. Would you mind creating a new PR for this? Sorry for the troubles.
